metrics/src/vespa/metrics/prometheus_writer.cpp

```diff
@@ -0,0 +1,287 @@
+}
+
+vespalib::string PrometheusWriter::escaped_label_value(stringref value) {
+    asciistream out;
+    for (char ch : value) {
+        if (ch == '\\') {
+            out << "\\\\";
+        } else if (ch == '"') {
+            out << "\\\"";
+        } else if (ch == '\n') {
+            out << "\\n";
+        } else [[likely]] {
+            out << ch; // assumed to be part of a valid UTF-8 sequence
+        }
+    }
+    return out.str();
+}
+
+stringref PrometheusWriter::stable_label_value_string_ref(stringref raw_label_value) {
+    if (!label_value_needs_escaping(raw_label_value)) [[likely]] {
+        return arena_stable_string_ref(raw_label_value);
+    } else {
+        return arena_stable_string_ref(escaped_label_value(raw_label_value));
+    }
+}
+
+void PrometheusWriter::build_labels_upto_root(vespalib::SmallVector<stringref, 16>& out, const Metric& m) {
+    const Metric* current = &m;
+    do {
+        for (const auto& tag : current->getTags()) {
+            if (!tag.hasValue()) {
+                continue; // Don't emit value-less tags, as these are not proper labels
+            }
+            out.emplace_back(stable_name_string_ref(tag.key()));
+            out.emplace_back(stable_label_value_string_ref(tag.value()));
+        }
+        current = current->getOwner();
+    } while (current != nullptr);
+}
+
+ConstArrayRef<stringref> PrometheusWriter::as_prometheus_labels(const Metric& m) {
+    if (!any_metric_in_path_has_nonempty_tag(m)) {
+        return {};
+    }
+    vespalib::SmallVector<stringref, 16> kv_refs;
+    build_labels_upto_root(kv_refs, m);
+    return _arena.copy_array<stringref>(kv_refs);
+}
+
+bool PrometheusWriter::visitSnapshot(const MetricSnapshot& ms) {
+    // Pre-cache timestamp in string form to avoid same conversion for every time series
+    _timestamp_str = std::to_string(std::chrono::duration_cast<std::chrono::milliseconds>(
+            ms.getToTime().time_since_epoch()).count());
+    return true;
+}
+
+void PrometheusWriter::doneVisitingSnapshot(const MetricSnapshot&) {
+    // No-op
+}
+
+bool PrometheusWriter::visitMetricSet(const MetricSet& set, bool) {
+    // Don't include metric sets that will be aggregated up into a separate sum metric.
+    // We don't care about individual threads etc., just their aggregate values.
+    if (set.hasTag("partofsum")) {
+        return false;
+    }
+    if (set.getOwner()) {
+        _path.emplace_back(stable_name_string_ref(set.getName()));
+    } // else: don't add the topmost set
+    return true;
+}
+
+void PrometheusWriter::doneVisitingMetricSet(const MetricSet& set) {
+    if (set.getOwner()) {
+        assert(!_path.empty());
+        _path.pop_back();
+    }
+}
+
+bool PrometheusWriter::visitCountMetric(const AbstractCountMetric& m, bool) {
+    auto full_path = metric_to_path_ref(m.getName());
+    auto labels    = as_prometheus_labels(m);
+    _samples.emplace_back(TimeSeriesSample{full_path, "count", labels, {m.getLongValue("count")}});
```

Comment:
Is the TimeSeriesSample needed.
Are there ambiguities preventing just emplace_back(full_path, "count", labels, {m.getLongValue("count")}) ?

Reply:
This was done due to avoid type deduction problems when `emplace_back`-forwarding a nested `{}` initializer of the `std::variant` argument, but in retrospect this isn't needed at all since just passing the value itself will implicitly construct the variant with the correct value. Will simplify.